internal/runner/runner.go

```diff
@@ -0,0 +1,121 @@
+package runner
+
+import (
+	"context"
+	"time"
+
+	"github.com/metal-toolbox/flasher/internal/model"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// A Runner instance runs a single task, setting up and executing the actions required to install firmware
+// on one or more server components.
+type Runner struct {
+	logger *logrus.Entry
+}
+
+type Handler interface {
+	Initialize(ctx context.Context) error
+	Query(ctx context.Context) error
+	PlanActions(ctx context.Context) error
+	RunActions(ctx context.Context) error
+	OnSuccess(ctx context.Context, task *model.Task)
+	OnFailure(ctx context.Context, task *model.Task)
+	Publish()
+}
+
+func New(logger *logrus.Entry) *Runner {
+	return &Runner{
+		logger: logger,
+	}
+}
+
+func (r *Runner) RunTask(ctx context.Context, task *model.Task, handler Handler) error {
+	// nolint:govet // struct field optimization not required
+	funcs := []struct {
```

Comment:
For future work, you might consider ways to abstract the "what" to do from the "how" to do it. I spent a fair amount of time working that sort of thing out at DO. We could pull that out and turn it into a module in `rivets` so we would have a common execution framework.